docs/RFCS/20200810_leaseholder_cache_invalidation.md

```diff
@@ -0,0 +1,464 @@
+
+## 3. Gossiping lease acquisitions
+
+Outline:
+
+- each node acquiring a lease (either because it was given to it
+  during another's node drain, or it is voted to be when a node goes
+  AWOL) gossips this fact in a "lease record", keyed by the range ID
+  with the lease as payload.
+- upon receiving "lease records", other nodes update their cache
+  with the gossiped lease, if it's newer than what they know.
+
+This approach was prototyped here: https://github.com/cockroachdb/cockroach/pull/52572
+
+**Pros:**
+
+- Addresses both [scenarios A and B](#scenarios).
+- By using gossip, the data will eventually reach every other
+  node even those that were temporarily unavailable.
+  (i.e. eliminates one cons of solution 1 above)
+- By informing the other nodes of the new leaseholder,
+  this solution accelerates/removes the need for the
+  lease discovery mechanism.
+  (i.e. eliminates the other cons of solution 1 above)
+
+**Cons:**
+
+- There is one "lease record" in gossip per range.
+  This violates the [design principles](#general-cockroachdb-design-principles).
+
+  We can partially alleviate this cost by populating the lease expiry
+  time as the value of [gossip TTL](#gossip). This reduces
+  the space complexity from O(#ranges) to O(#active ranges).
+
+- Like above, there's a delay between the moment a node terminates
+  ([step 5 of drain](#draining)) and the moment other nodes learn of
+  the new lease records. During that time, requests can be mis-routed.
+
+  Like above, this last cons can be alleviated by extending step 5 of
+  the draining process by a "sleep period".
+
+### Multi-tenant variant
+
+- The KV servers receive the "lease records" over gossip from other KV servers.
+- The KV servers *forward* the lease records to SQL tenants,
+  where the range caches are stored.
+
+## 4. Liveness-based cache invalidation with poison
+
+Outline:
+
+- [liveness updates](#liveness-updates) are gossiped as usual.
+- upon receiving a liveness update that indicates that another node is
+  not live, each node updates its range cache as follows:
+
+  - invalidate all leases pointing to the now-unavailable node, ie.
+    remove the lease from the cache.
+  - mark that store/node ID as "not preferred" for the subsequent lease
+    discovery. This *“poisons”* that store/node ID.
+
+- upon a cache miss, the range cache will still do lease discovery,
+  but it will skip over "not preferred" nodes/stores in the first
+  round.
+
+  Only if a lease is not discovered during the first round, are
+  poisoned nodes added to the lookup in the second and next rounds.
+
+**Pros:**
+
+- Addresses both [scenarios A and B](#scenarios).
+- By using gossip, the data will eventually reach every other
+  node even those that were temporarily unavailable.
+  (i.e. eliminates one cons of solution 1 above)
+- Does not add new data in gossip. Does not violate the [design
+  principles](#general-cockroachdb-design-principles).
+
+**Cons:**
```

Comment:
the biggest con here, in my opinion, is that this doesn't help avoid bad routing on regular lease transfers; it just helps with nodes restarting or losing liveness or such.